File: src/main/java/pl/mroz/buddiesapi/infrastructure/database/rental/RentalRepositoryImpl.java
package pl.mroz.buddiesapi.infrastructure.database.rental;

import lombok.RequiredArgsConstructor;
import org.apache.commons.lang3.NotImplementedException;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.JpaSpecificationExecutor;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import pl.mroz.buddiesapi.domain.rental.Rental;
import pl.mroz.buddiesapi.domain.rental.RentalRepository;

import java.util.List;
import java.util.UUID;

@RequiredArgsConstructor
public class RentalRepositoryImpl implements RentalRepository {

    private final RentalJpaRepository jpaRepository;

    @Override
    public List<Rental> getAllRentals() {
        return jpaRepository.findAll()
                .stream()
                .map(Rental::fromDb)
                .toList();
    }

    @Override
    public Page<Rental> getByCriteria(Specification<RentalEntity> specification, Pageable pageable) {
        return jpaRepository.findAll(specification, pageable).map(Rental::fromDb);
    }

    @Override
    public Rental getRentalById(UUID rentalId) {
        return jpaRepository.findById(rentalId)
                .map(Rental::fromDb)
                .orElseThrow(
                        () -> new IllegalArgumentException("No Rental with UUID: " + rentalId)
                );
    }

    @Override
    public List<Rental> getAllRentalsFromUser(UUID accountId) {
        return jpaRepository.getAllByAuthorEntityId(accountId)
                .stream()
                .map(Rental::fromDb)
                .toList();
    }

    @Override
    public List<String> getTopTags() {
        throw new NotImplementedException();
//        return jpaRepository.getTopTags();
    }

    @Override
    public Rental save(Rental rental) {
        var savedEntity = jpaRepository.save(RentalEntity.fromDomain(rental));
        return Rental.fromDb(savedEntity);
    }

    @Override
    public void delete(Rental rental) {
        jpaRepository.findById(rental.getRentalId())
                .ifPresent(jpaRepository::delete);
    }

    interface RentalJpaRepository extends JpaRepository<RentalEntity, UUID>, JpaSpecificationExecutor<RentalEntity> {

//        @Query(value = "SELECT feature_tags FROM public.feature_tags Group By feature_tags Order By COUNT(*) DESC LIMIT 5",
//                nativeQuery = true)
//        List<String> getTopTags();

        @Query(value = "SELECT r FROM RentalEntity r where r.authorEntity.accountId = :authorId")
        List<RentalEntity> getAllByAuthorEntityId(@Param("authorId") UUID authorId);

    }
}

File: src/main/kotlin/buddiesapi/infrastructure/database/rental/RentalRepositoryImpl.kt
package buddiesapi.infrastructure.database.rental

import buddiesapi.domain.rental.Rental
import buddiesapi.domain.rental.RentalRepository
import buddiesapi.infrastructure.database.rental.RentalEntity.Companion.fromDomain
import java.util.UUID
import mu.KLogging
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.jpa.domain.Specification
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.JpaSpecificationExecutor
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param

class RentalRepositoryImpl(
    private val jpaRepository: RentalJpaRepository
) : RentalRepository {

    override fun getAllRentals(): List<Rental> = jpaRepository.findAll().stream()
        .map { entity: RentalEntity? -> entity?.let { Rental.fromDb(it) } }
        .toList()

    override fun getByCriteria(criteria: Specification<RentalEntity>, pageable: Pageable): Page<Rental> =
        jpaRepository.findAll(criteria, pageable)
            .map { entity: RentalEntity? ->
                entity?.let { Rental.fromDb(it) }
            }

    override fun getRentalById(rentalId: UUID): Rental? =
        jpaRepository.getByRentalId(rentalId)?.let {
            Rental.fromDb(it)
        }

    override fun getAllRentalsFromUser(accountId: UUID): List<Rental> =
        jpaRepository.getAllByAuthorEntityId(accountId)
            .stream()
            .map { entity: RentalEntity? ->
                entity?.let { Rental.fromDb(it) }
            }
            .toList()


    override val topTags: List<String> = emptyList() //todo

    override fun save(rental: Rental): Rental {
        logger.info { "Saving ${rental.rentalId}, with userID: ${rental.author.accountId}" }
        val savedEntity = jpaRepository.save(fromDomain(rental))
        return Rental.fromDb(savedEntity)
    }

    override fun delete(rental: Rental) =
        jpaRepository.delete(fromDomain(rental))


    companion object : KLogging()

    interface RentalJpaRepository : JpaRepository<RentalEntity?, UUID>,
        JpaSpecificationExecutor<RentalEntity> {
        //        @Query(value = "SELECT feature_tags FROM public.feature_tags Group By feature_tags Order By COUNT(*) DESC LIMIT 5",
        //                nativeQuery = true)
        //        List<String> getTopTags();
        @Query(value = "SELECT r FROM RentalEntity r where r.authorEntity.accountId = :authorId")
        fun getAllByAuthorEntityId(@Param("authorId") authorId: UUID): List<RentalEntity>

        fun getByRentalId(accountId: UUID): RentalEntity?
    }
}
